### PR TITLE
http_client: ssl context optimization

### PR DESCRIPTION
**Pull request by abhilashraju**
ssl context are recreated whenever new ssl stream is created. We can reuse the already created ssl context object for any new ssl streams. This will reduce the number of ssl context object creation in the negative retry scenarios.
We have seen libcrypto crashes in the retry forever scenario. Reducing the number of ssl context object creation  may fix such crashes.

Tested By:
Following unit tests has been executed.
1) Basic sanity tests on eventing
2) Negative stale subscription cases

Need to execute all system level cases convering HMC/BMC connetion issues before merging

### CODE REVIEW

File: http/http_client.hpp [hunk 1]
```diff
@@ -468,7 +467,7 @@ class ConnectionInfo : public std::enable_shared_from_this<ConnectionInfo>
             }
             if (connPolicy->retryPolicyAction == "RetryForever")
             {
-                state = ConnState::idle;
+                state = ConnState::closed;
```
Comment (gtmills) [inline http/http_client.hpp:470]: Why closed over terminate or suspend?

**Comment by baemyung** [comment]
Isn't it good to go thru upstream?

>We have seen libcrypto crashes in the retry forever scenario. Reducing the number of ssl context object creation may fix such crashes.

Is there a root cause analysis what causes the crash?
If not, how would this fix the crashes although the crashes may be less often?

File: http/http_client.hpp (hunk 1 above)
Comment (raviteja-b) [reply http/http_client.hpp:470]: @abhilashraju have we tested this change in retry forever scenario?
looks like this case needs to be tested.

Comment (abhilashraju) [reply http/http_client.hpp:470]: @raviteja-b yes I have done retry test . Doesnot show any issues in dev testing.
@gtmills  the  callback(false, connId, res); at 476  will set the connection state to closed anyway. do we really need this line of code. I don't know the history. Can we remove this code to simplify ?

**Comment by abhilashraju** [comment]
> Isn't it good to go thru upstream?
> 
> > We have seen libcrypto crashes in the retry forever scenario. Reducing the number of ssl context object creation may fix such crashes.
> 
> Is there a root cause analysis what causes the crash? If not, how would this fix the crashes although the crashes may be less often?

We could reproduce it only once. The callstack points to libcrypto. At the moment we can only speculate the root cause. As I said in the commitlog, it need not fix the issue. But still it is an optimisation we can do. Yes we can propose the same to upstream. But the context creation is done a bit differently in upstream. We have commented it out in downstream. I don't know the reason.

**Comment by gtmills** [comment]
What is our plan here? 
Should we close and bring back if we think we need in 1060?

**Pull request closed without merging**